### PR TITLE
Do not try to encode strings passed to the str() constructor (#1108)

### PR DESCRIPTION
**Pull request by kunom**
See issue #1108.

### CODE REVIEW

**Comment by kunom** [comment]
Passes `test-ipy.cmd /test:test_str_20 /test:test_unicode_20 /test:test_str_bytes_20`.

**Comment by simplicbe** [comment]
LGTM, @paweljasinski what do you think? Thank you very much @kunom!

**Comment by kunom** [comment]
@simplicbe Yes. Once merged, you can close #1108.

File: Languages/IronPython/IronPython/Runtime/Operations/StringOps.cs [hunk 1]
```diff
@@ -199,9 +199,10 @@ public static class StringOps {
             if (x == null) {
                 return "None";
             }
-            if (x is string) {
-                // check ascii
-                return CheckAsciiString(context, (string)x);
+
+            string xstr = (x as string);
```
Comment (slide) [inline Languages/IronPython/IronPython/Runtime/Operations/StringOps.cs]: Do we need to check Extensible<string> here too?

Comment (kunom) [inline Languages/IronPython/IronPython/Runtime/Operations/StringOps.cs]: I don't think so:
```
Python 2.7.12 (v2.7.12:d33e0cf91556, Jun 27 2016, 15:19:22) [MSC v.1500 32 bit (Intel)] on win32
>>> class S(str): pass
...
>>> a = S('foo')
>>> b = str(a)
>>> a is b
False
>>> c = str(b)
>>> c is b
True
```

If the check in `FastNew` would also include `Extensible<string>` then the first check in the sample code would evaluate to `True`.

**Comment by paweljasinski** [comment]
I have put together a test and run under cpython, ipy with this PR and jython.
```
# -*- coding: utf-8 -*-
s = str("ä")
print len(s)
for c in s:
    print "%x" % ord(c)

print
s = str("\xc3\xa4")
print len(s)
for c in s:
    print "%x" % ord(c)

print
class A(str):
    pass
s = A("ä")
print len(s)
for c in s:
    print "%x" % ord(c)

# throws under cpython and jython, ok under ipy
s = str(u"ä")
```

cpython
```
$ python /home/rejap/tmp/str2.py
2
c3
a4

2
c3
a4

2
c3
a4
Traceback (most recent call last):
  File "/home/rejap/tmp/str2.py", line 22, in <module>
    s = str(u"ä")
UnicodeEncodeError: 'ascii' codec can't encode character u'\xe4' in position 0: ordinal not in range(128)
```

ipy
```
1
e4

2
c3
a4

1
e4
```

jython
```
2
c3
a4

2
c3
a4

2
c3
a4
Traceback (most recent call last):
  File "c:/cygwin64/home/rejap/tmp/str2.py", line 22, in <module>
    s = str(u"\xc3\xa4")
UnicodeEncodeError: 'ascii' codec can't encode character u'\xe4' in position 0: ordinal not in range(128)
```

even though the results are not consistent, particularly missing exception in our case. I have a feeling that merging would improve the situation.
At the same time, jython appears to address it better than ipy does, so probably things can be done better.

**Comment by kunom** [comment]
@paweljasinski thanks.

IronPython took the decision that `str` and `unicode` are basically two aliases of the same underlying type which is called `unicode` in Python2 (but `str` in Python3). As long as `type('a') == type(u'a')` evaluates to `true` (Python3 and IronPython behaviour), you will never reach full Python2 compatibility.

**Comment by paweljasinski** [comment]
I still wonder how is it done in jython. They work with the underlying jvm with unicode strings as well and somehow avoid problems we experience.

**Comment by kunom** [comment]
@paweljasinski I don't know jython and its coupling to the JVM at all, but from the REPL frontend, they just seem to support both byte strings as well as unicode strings as different things:

```
Jython 2.7.0 (default:9987c746f838, Apr 29 2015, 02:25:11)
[Java HotSpot(TM) 64-Bit Server VM (Oracle Corporation)] on java1.8.0_60
Type "help", "copyright", "credits" or "license" for more information.
>>> type('a')
<type 'str'>
>>> type(u'a')
<type 'unicode'>
>>> ord('õ')
228
>>> ord(u'õ')
245
```

An excerpt from the [jython documentation](http://www.jython.org/jythonbook/en/1.0/DataTypes.html#strings-and-string-methods):

> Prior to the release of Jython 2.5.0 there was only one string type. The string type in Jython supported full two-byte Unicode characters and all functions contained in the string module were Unicode-aware. If the u’’ string modifier is specified, it is ignored by Jython. Since the release of 2.5.0, strings in Jython are treated just like those in CPython, so the same rules will apply to both implementations.

Separating _str_ from _unicode_ in IronPython would certainly increase CPython compatibility, but at the same time, much more thinking must go into interoperability with other .NET code. It would be interesting to know how Jython approaches that problem.

However, I think, that discussion is far outside of the current pull request that only tries to make the string/unicode constructor that receives a string/unicode argument being a null operation.

**Comment by slide** [comment]
@paweljasinski are you happy with this? we can look into what jython did if we are interested, but I think this is pretty self contained for now.

**Comment by paweljasinski** [comment]
I am happy with the fix.

**Pull request merged**